django/db/models/expressions.py

```diff
@@ -483,7 +485,14 @@ def __init__(self, value, output_field=None):
         self.value = value
 
     def as_sql(self, compiler, connection):
-        return '%s', [self.value]
+        val = self.value
+        if self._output_field_or_none is not None:
+            from django.db.models.sql.compiler import SQLUpdateCompiler
+            if isinstance(compiler, SQLUpdateCompiler):
```

Comment:
OK I've had a play around, and these are my findings so far.

Value.as_sql:

```
def as_sql(self, compiler, connection):
        val = self.value
        if self._output_field_or_none is not None:
            val = self.output_field.get_db_prep_value(val, connection=connection)
        return '%s', [val]
```

This appears to work fine for all backends. No need to differentiate between get_db_prep_{save/value} that I can find. Were there any particular examples that required you to use one but not the other?

BaseCaseExpression.as_postgresql:

- I got rid of the as_postgresql completely, and there are failing tests. The tests that fail are types that look like strings, but represent things like Binary or IPAddress.

- I got rid of the check for Update, to apply the cast to all queries. It fails on CharFields (because varchar(None) isn't a valid type. This highlighted another issue - if we were updating CharFields and the Values have an output_field, then it fails with the same error of varchar(None):

```
CaseTestModel.objects.filter(string__in=['1', '2']).update(
            string=self.create_expression(
                'integer',
                [(Value(1), Value('1', output_field=models.CharField())),
                 (Value(2), Value('2', output_field=models.CharField()))]))
```

Any datatype that requires extra information, like max_length or even max_digits for DecimalField will see similar issues.

What I think we need to do for the BaseCaseExpression.as_postgresql, is to apply casting only to fields that require it, and to do it whether it's a select or an update (do not discriminate). 

Maybe we can get get_db_prep to ask the backend to provide casting info for fields that require it - like UUID or GenericIPAddress. In the short term maybe we could document certain fields as not supported with certain expressions.

But this is scope creep, and should probably be handled or discussed separately. If anyone has any input here it would be helpful. If I'm not being clear please speak up too.